File: frontend/src/lib/api/intelligence.ts
import { apiClient, toApiClientError } from "@/lib/api/client";
import type { ApiResponse } from "@/lib/api/types";

/** GET /api/v1/intelligence/adaptive-weights — Adaptive Scoring Weights
 * (Task 1/7, Adaptive Intelligence round): the same multipliers the
 * backend's own compute_lead_score() already applies to its matching
 * bonuses (industry:X, company_size:X, action:X, fast_response,
 * high_risk), exposed here purely for visibility. Keys with no real
 * 30-day signal yet are simply absent, not defaulted to 1.0. */
export async function getAdaptiveWeights(): Promise<Record<string, number>> {
  try {
    const { data } = await apiClient.get<ApiResponse<Record<string, number>>>(
      "/intelligence/adaptive-weights"
    );
    return data.data ?? {};
  } catch (error) {
    throw toApiClientError(error);
  }
}

export interface RevenueSimulation {
  currentExpected: number;
  optimizedExpected: number;
  delta: number;
}

interface RevenueSimulationDto {
  current_expected: number;
  optimized_expected: number;
  delta: number;
}

/** GET /api/v1/intelligence/revenue-simulation — Revenue Simulation Engine
 * (Task 5/7): a deterministic "best case" projection assuming every
 * pending action gets executed, not a forecast. */
export async function getRevenueSimulation(): Promise<RevenueSimulation> {
  try {
    const { data } = await apiClient.get<ApiResponse<RevenueSimulationDto>>(
      "/intelligence/revenue-simulation"
    );
    return {
      currentExpected: data.data?.current_expected ?? 0,
      optimizedExpected: data.data?.optimized_expected ?? 0,
      delta: data.data?.delta ?? 0,
    };
  } catch (error) {
    throw toApiClientError(error);
  }
}

/** GET /api/v1/intelligence/exec-insight — CEO Insight Layer (Task 6/7):
 * one ready-to-render sentence combining today's lost opportunity with
 * the Revenue Simulation Engine's own optimistic delta. */
export async function getExecInsight(): Promise<string> {
  try {
    const { data } = await apiClient.get<ApiResponse<{ message: string }>>(
      "/intelligence/exec-insight"
    );
    return data.data?.message ?? "";
  } catch (error) {
    throw toApiClientError(error);
  }
}

export type GlobalStrategyFocus = "calls" | "messages" | "meetings";

export interface GlobalStrategy {
  focus: GlobalStrategyFocus;
  reason: string;
  confidence: number;
}

interface GlobalStrategyDto {
  focus: GlobalStrategyFocus;
  reason: string;
  confidence: number;
}

/** GET /api/v1/intelligence/global-strategy — Global Strategy Engine
 * (final round, Task 2/8): one recommended channel to focus on right now
 * (calls/messages/meetings), plain deterministic rules. */
export async function getGlobalStrategy(): Promise<GlobalStrategy | null> {
  try {
    const { data } = await apiClient.get<ApiResponse<GlobalStrategyDto>>(
      "/intelligence/global-strategy"
    );
    if (!data.data) return null;
    return { focus: data.data.focus, reason: data.data.reason, confidence: data.data.confidence };
  } catch (error) {
    throw toApiClientError(error);
  }
}

export type AggressionLevel = "low" | "medium" | "high" | "extreme";
/** compute_revenue_mode()'s own relabeling of AggressionLevel (final
 * round, Task 6/8) — same underlying signal, "efficiency"/"balanced"/
 * "aggressive" vocabulary instead, for the Command Center's own Revenue
 * Mode indicator. */
export type RevenueMode = "efficiency" | "balanced" | "aggressive";

export interface AggressionState {
  level: AggressionLevel;
  revenueMode: RevenueMode;
}

/** GET /api/v1/intelligence/aggression-level — Dynamic Aggression Mode
 * (final round, Task 3/8): how hard the system should be pushing right
 * now, derived from today's lost opportunity and the Revenue Simulation
 * Engine's own gap, plus its own revenue_mode relabeling. */
export async function getAggressionLevel(): Promise<AggressionState | null> {
  try {
    const { data } = await apiClient.get<
      ApiResponse<{ level: AggressionLevel; revenue_mode: RevenueMode }>
    >("/intelligence/aggression-level");
    if (!data.data) return null;
    return { level: data.data.level, revenueMode: data.data.revenue_mode };
  } catch (error) {
    throw toApiClientError(error);
  }
}

export interface RevenueLeaks {
  leadsIgnoredOver24h: number;
  highValueLeadsWithoutAction: number;
  leadsStuckSameStage: number;
  totalLeakValue: number;
}

interface RevenueLeaksDto {
  leads_ignored_over_24h: number;
  high_value_leads_without_action: number;
  leads_stuck_same_stage: number;
  total_leak_value: number;
}

/** GET /api/v1/intelligence/revenue-leaks — Revenue Leak Detector (final
 * round, Task 5/8): three independent leak patterns plus totalLeakValue,
 * the summed expectedValue across every lead caught by at least one. */
export async function getRevenueLeaks(): Promise<RevenueLeaks> {
  try {
    const { data } = await apiClient.get<ApiResponse<RevenueLeaksDto>>(
      "/intelligence/revenue-leaks"
    );
    return {
      leadsIgnoredOver24h: data.data?.leads_ignored_over_24h ?? 0,
      highValueLeadsWithoutAction: data.data?.high_value_leads_without_action ?? 0,
      leadsStuckSameStage: data.data?.leads_stuck_same_stage ?? 0,
      totalLeakValue: data.data?.total_leak_value ?? 0,
    };
  } catch (error) {
    throw toApiClientError(error);
  }
}

File: backend/app/schemas/intelligence.py
from pydantic import BaseModel


class RevenueSimulationResponse(BaseModel):
    """GET /intelligence/revenue-simulation — simulate_revenue_if_all_
    actions_executed()'s own {current_expected, optimized_expected, delta}
    shape (scoring.py), typed for the API contract. current_expected is
    the same win_probability-weighted sum score_leads() already exposes
    per lead (LeadResponse.expected_value); optimized_expected assumes
    every pending action gets executed and every open lead's own
    win_probability is boosted accordingly — see that function's own
    docstring for the exact, disclosed assumption. delta is always
    optimized_expected - current_expected."""

    current_expected: int = 0
    optimized_expected: int = 0
    delta: int = 0


class ExecInsightResponse(BaseModel):
    """GET /intelligence/exec-insight — generate_exec_insight()'s own
    ready-to-render sentence (app/services/leads/intelligence.py),
    combining today's lost opportunity, the Revenue Simulation Engine's
    own optimistic delta, the current aggression level, the global
    strategy's focus, and any revenue leak count (final round). Same "the
    backend writes the sentence" rule build_priority_reason()/
    focus_message/accountability_message already follow elsewhere in this
    codebase."""

    message: str


class GlobalStrategyResponse(BaseModel):
    """GET /intelligence/global-strategy — compute_global_strategy()'s own
    {focus, reason, confidence} shape (scoring.py), typed for the API
    contract. focus is always one of "calls"/"messages"/"meetings";
    confidence is 0-100, not a probability in the statistical sense — a
    plain "how strong is this rule's own signal" indicator."""

    focus: str
    reason: str
    confidence: int = 0


class AggressionLevelResponse(BaseModel):
    """GET /intelligence/aggression-level — compute_aggression_level()'s
    own return value (scoring.py), one of "low"/"medium"/"high"/
    "extreme". revenue_mode (final round) is compute_revenue_mode()'s own
    one-line relabeling of that same level into "efficiency"/"balanced"/
    "aggressive" — the vocabulary this round's own Command Center
    indicator asks for, same underlying signal as `level`, not a second
    computation (see that function's own docstring)."""

    level: str
    revenue_mode: str = "balanced"


class RevenueLeaksResponse(BaseModel):
    """GET /intelligence/revenue-leaks — detect_revenue_leaks()'s own
    shape (scoring.py): three independent leak-pattern counts plus
    total_leak_value, the summed expected_value across every lead caught
    by at least one of the three patterns (no double-counting a lead
    caught by more than one)."""

    leads_ignored_over_24h: int = 0
    high_value_leads_without_action: int = 0
    leads_stuck_same_stage: int = 0
    total_leak_value: int = 0
